File: Helpers/src/main/java/com/dg/spannable/CustomTypefaceSpannable.java
package com.dg.spannable;

import android.graphics.Paint;
import android.graphics.Typeface;
import android.text.TextPaint;
import android.text.style.MetricAffectingSpan;

/**
 * Created by Daniel Cohen Gindi (danielgindi@gmail.com)
 *
 * Changes the typeface family of the text to which the span is attached.
 */
public class CustomTypefaceSpannable extends MetricAffectingSpan
{
    private Typeface mTypeface;

    private static int SPANNABLE_ID = 100;

    public CustomTypefaceSpannable(Typeface typeface)
    {
        mTypeface = typeface;
    }

    public int getSpanTypeId()
    {
        return SPANNABLE_ID;
    }

    public int describeContents()
    {
        return 0;
    }

    public Typeface getTypeface()
    {
        return mTypeface;
    }

    @Override
    public void updateDrawState(TextPaint ds)
    {
        apply(ds, mTypeface);
    }

    @Override
    public void updateMeasureState(TextPaint paint)
    {
        apply(paint, mTypeface);
    }

    private static void apply(Paint paint, Typeface typeface)
    {
        paint.setTypeface(typeface);
    }
}

File: Helpers/src/main/java/com/dg/spannable/CustomTypefaceSpannable.kt
package com.dg.spannable

import android.graphics.Paint
import android.graphics.Typeface
import android.text.TextPaint
import android.text.style.MetricAffectingSpan

/**
 * Changes the typeface family of the text to which the span is attached.
 */
@Suppress("unused")
class CustomTypefaceSpannable : MetricAffectingSpan
{
    @Suppress("MemberVisibilityCanBePrivate")
    val typeface: Typeface

    @Suppress("ConvertSecondaryConstructorToPrimary")
    constructor(typeface: Typeface) : super()
    {
        this.typeface = typeface
    }

    val spanTypeId: Int
        get() = SPANNABLE_ID

    fun describeContents(): Int
    {
        return 0
    }

    override fun updateDrawState(ds: TextPaint)
    {
        apply(ds, typeface)
    }

    override fun updateMeasureState(paint: TextPaint)
    {
        apply(paint, typeface)
    }

    companion object
    {
        private const val SPANNABLE_ID = 100

        private fun apply(paint: Paint, typeface: Typeface)
        {
            paint.typeface = typeface
        }
    }
}